### PR TITLE
[RV64_DYNAREC] Added more support for XTheadMemPair extension

### PR DESCRIPTION
**Pull request by ksco**

### CODE REVIEW

File: src/dynarec/rv64/dynarec_rv64_helper.h [hunk 1]
```diff
@@ -830,206 +947,208 @@
...
-#define emit_shrd32c    STEPNAME(emit_shrd32c)
-#define emit_shld32c    STEPNAME(emit_shld32c)
-
-#define emit_pf         STEPNAME(emit_pf)
-
-#define x87_do_push     STEPNAME(x87_do_push)
+#define native_pass STEPNAME(native_pass)
+
+#define dynarec64_00 STEPNAME(dynarec64_00)
+#define dynarec64_00_0 STEPNAME(dynarec64_00_0)
```
Comment (ptitSeb) [inline src/dynarec/rv64/dynarec_rv64_helper.h:1002]: meh, alignent is gone :(

File: src/dynarec/rv64/rv64_emitter.h [hunk 2]
```diff
@@ -1029,11 +1029,29 @@ f28–31  ft8–11  FP temporaries                  Caller
...
 
-// TODO
-// th.lwd rd1, rd2, (rs1), imm2, 3 Load two signed 32-bit values
-// th.lwud rd1, rd2, (rs1), imm2, 3 Load two unsigned 32-bit values
-// th.sdd rd1, rd2, (rs1), imm2, 4 Store two 64-bit values
-// th.swd rd1, rd2, (rs1), imm2, 3 Store two 32-bit values
+// Load two signed 32-bit values from memory into two GPRs.
+// addr := rs1 + (zero_extend(imm2) << 3)
+// reg[rd1] := sign_extend(mem[addr+3:addr])
+// reg[rd2] := sign_extend(mem[addr+7:addr+3])
```
Comment (ptitSeb) [inline src/dynarec/rv64/rv64_emitter.h:1035]: // reg[rd2] := sign_extend(mem[addr+7:addr+4])

File: src/dynarec/rv64/dynarec_rv64_helper.h (hunk 1 above)
Comment (ksco) [reply src/dynarec/rv64/dynarec_rv64_helper.h:1002]: Alignments are back!

File: src/dynarec/rv64/rv64_emitter.h [hunk 3]
```diff
@@ -1029,11 +1029,29 @@ f28–31  ft8–11  FP temporaries                  Caller
...
+#define TH_LWD(rd1, rd2, rs1, imm2) EMIT(R_type(0b1110000 | ((imm2)&0b11), rd2, rs1, 0b100, rd1, 0b0001011))
+
+// Load two unsigned 32-bit values from memory into two GPRs.
+// addr := rs1 + (zero_extend(imm2) << 3)
+// reg[rd1] := zero_extend(mem[addr+3:addr])
+// reg[rd2] := zero_extend(mem[addr+7:addr+3])
```
Comment (ptitSeb) [inline src/dynarec/rv64/rv64_emitter.h:1041]: same

File: src/dynarec/rv64/rv64_emitter.h (hunk 2 above)
Comment (ksco) [reply src/dynarec/rv64/rv64_emitter.h:1035]: wow, this is interesting, I'm actually copied these pseudo codes directly from the thead spec.

Comment (ksco) [reply src/dynarec/rv64/rv64_emitter.h]: Fixed, I'll submit a PR to their spec repo.

Comment (ptitSeb) [reply src/dynarec/rv64/rv64_emitter.h]: I still don't think it's `[addr+7:addr+3]` interval, that would be 36bits. Or mayb it's `[addr+7:addr+3[` but that's not ass clear as `[addr+7:addr+4]`

Comment (ksco) [reply src/dynarec/rv64/rv64_emitter.h]: Yeah, I think it's a typo, which is ugh... anyway...

Comment (ksco) [reply src/dynarec/rv64/rv64_emitter.h]: > Fixed, I'll submit a PR to their spec repo.

Done: https://github.com/T-head-Semi/thead-extension-spec/pull/29

**Pull request merged**